index.rst

```diff
@@ -1,29 +1,481 @@
+       | each dataset (except log and metadata datasets).
+   * - ``logs``
+     - multi-block
+     - The content of all butler log datasets (one dataset per block).
+   * - ``metadata``
+     - multi-block
+     - The content of all butler metadata datasets (one dataset per block).
+   * - ``quantum_addresses``
+     - address (arbitrary order)
+     - | UUID to integer ID mapping for all quanta; byte offsets and sizes
+       | for the ``quanta``, ``logs``, and ``metadata`` components.
+   * - ``dataset_addresses``
+     - address (arbitrary order)
+     - | UUID to integer ID mapping for all datasets; byte offsets and sizes
+       | for the ``datasets`` component.
+
+When the aggregation graph is first constructed, the ``header`` and ``bipartite_edges`` components are constructed in full from the information in the predicted or wire graph immediately.
+The other components are appended to as quanta are aggregated:
+
+1. When a set of quanta is recognized as having finished (successfully or unsuccessfully), either by receipt of a message or by scanning for the metadata and log datasets of unblocked quanta, their predicted-quantum information is loaded from the original predicted or wire graph.
+2. When metadata datasets exist (i.e. the quantum was successful), they are read in full to retrieve status information and dataset provenance.  For quanta that failed (the log exists but metadata does not, or a flag from the user indicates that all unsuccessful quanta should now be considered failures), the existence of all predicted output datasets is checked.
+3. Information about output datasets (including those that were predicted but not produced, but not including metadata and logs) is appended to the ``datasets`` multi-block file.
+4. The metadata datasets are reconverted to JSON, compressed, and appended to the ``metadata`` multi-block file.
+5. The log datasets are loaded, rewritten to JSON, compressed, and appended to the ``log`` multi-block file.
+6. If any quantum's metadata and log datasets are not inputs to any downstream quantum, they are deleted from their original location.  Similarly, if any of these quantum are the last consumers of an upstream metadata or log dataset, those upstream datasets are deleted from their original locations.
+7. Status and other per-quantum provenance is appended to the ``quanta`` multi-block file.
+8. Existing output datasets (not including metadata and logs) are ingested into the butler database.
+9. Addresses into the ``datasets`` multi-block file are appended to the ``dataset_addresses`` file.
+10. Addresses into the ``quanta``, ``metadata``, and ``logs`` multi-block files are appended to the ``quantum_addresses`` file.
+
+Most of the above steps can happen on many quanta in parallel (scanning, dataset reads, generating and compressing JSON), with a sequence point at the end to perform all file append, delete, and database insert operations in bulk from a single process/thread.
```

Comment:
You mean that at these sequence points a single batch submission can only have one process talking to the database "live"? But in theory multiple batch jobs could be hitting at once? Because this clashing is rare we do not need to consider kafka to control rate but might still have to put the aggregation job on its own queue to limit the number of parallelized butler writes?

Reply:
Are we aggregating logs/metadata/configs separately so that the finalJob zipping is fast? (although down below I see that maybe these datasets are not being stored at all in the future?)